### PR TITLE
Fault Tolerant Manual: Add support for DDP

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for Fault Tolerant Manual with DDP.

Parts of #10431

To be merged after
* #10568

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2094,7 +2094,11 @@ def _results(self) -> Optional[ResultCollection]:
             return active_loop._results
 
     def _exit_gracefully_on_signal(self) -> None:
-        if _fault_tolerant_training() and self._terminate_gracefully:
+        if _fault_tolerant_training():
+            # the signal should be sent to rank 0
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:2098]: are you sure? if the signal does not get sent to all processes, only some of them will enter this broadcast call and the rest will hang (they will never run the _exit_gracefully_on_signal method).

File: pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -737,3 +739,15 @@ def _teardown_dataloader_get_iterators() -> None:
     if get_iterator:
         DataLoader._get_iterator = get_iterator
         del DataLoader._ori_get_iterator
+
+
+def _collect_states_on_rank_zero_over_collection(state_dict: Any, device: torch.device) -> Any:
+    """This utility collects the state across processes for a collection of state."""
+
+    def fn(state: Dict):
+        nonlocal device
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:748]: this should not be necessary. the closure should bind the local variable

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -737,3 +739,15 @@ def _teardown_dataloader_get_iterators() -> None:
...
+        if state.get("state", None) is not None:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:749]:
```suggestion
        if "state" in state:
```
 or not?

Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:749]: I believe the function should document that it will only collect dictionaries with a key "state" in it.

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:2098]: This code works only if the signal is sent to global rank 0. I didn't manage to make it work otherwise. This would need more investigation.

File: pytorch_lightning/utilities/auto_restart.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:748]: removed !

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:749]: Done !

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 4]
```diff
@@ -174,11 +181,20 @@ def on_save_checkpoint(self) -> Dict:
...
     def on_load_checkpoint(self, state_dict: Dict) -> None:
         # cache the dataloader state dict until the dataloader objects are available
-        self._dataloader_state_dict = state_dict.get("dataloader_state_dict")
+        # dataset states are collected across all ranks
+        if _fault_tolerant_training():
+            dataloader_state_dict = state_dict.get("dataloader_state_dict", None)
+            if not dataloader_state_dict:
+                return
+            rank = torch.distributed.get_rank() if distributed_available() else 0
+            self._dataloader_state_dict = dataloader_state_dict[rank]
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: isn't it possible to do this before we call the `on_load_checkpoint` hook? I believe this shouldn't be only loop specific

Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: Yes, I was thinking about this too. And I believed the cleanest was there. 

If you do it within the base loop, the user won't be its full state_dict and this logic is hidden from the user.

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: I wouldn't do it in the base loop, but in the checkpoint connector on the full state dict, before we send it to the objects to restore. This line here

`self._dataloader_state_dict = dataloader_state_dict[rank]
`
tells me we can filter the dict already earlier before giving it to the objects. Isn't this better?

Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: I am afraid this would break the paradigm where state_dict should not being altered through connecting components such as the CheckpointConnector. I am personally not in favour of this and prefer to have the logic visible the reader.

This might also result in unwanted behaviour for users subclassing the loops and overriding those functions. How do you detect and automate the extraction of rank state dict in this case ?

@carmocca @ananthsub @justusschock Any thoughts ?

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: I think it's fine to have the loop call the collect function, but the rest should be made strategy agnostic as much as possible. For example, instead of `torch.distributed.get_rank()` use call Lightning to determine the global_rank.

File: pytorch_lightning/trainer/supporters.py [hunk 5]
```diff
@@ -403,6 +404,10 @@ def create_loader_iters(dataloader: DataLoader, state_dict: Dict) -> Iterator:
             if isinstance(dataloader, CycleIterator):
                 dataloader = dataloader_to_iter_on.loader
 
+            # dataset states are collected across all ranks
+            rank = torch.distributed.get_rank() if distributed_available() else 0
```
Comment (awaelchli) [inline pytorch_lightning/trainer/supporters.py:408]: alternatively, we could access Lightnings global rank variable, i.e., `_get_rank()` to make it more strategy agnostic, although granted it is not much cleaner.

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:748]: I believed you missed my comment here.
You also changed `_collect_states_on_rank_zero` so this code might be outdated.

File: pytorch_lightning/utilities/distributed.py [hunk 6]
```diff
@@ -378,7 +378,8 @@ def init_dist_connection(
         )
 
 
-def _collect_states_on_rank_zero(state: Dict[str, Any], device: torch.device) -> Optional[Dict[int, Any]]:
+# TODO: Refactor with the Strategy Collectives once finalized.
+def _collect_states_on_rank_zero(state: Dict[str, Any]) -> Optional[Dict[int, Any]]:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/distributed.py:382]: I believe this cannot return None anymore?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py:382]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:748]: Done !

File: pytorch_lightning/trainer/supporters.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py:408]: Yes, I believe we should do this when we have a cleaner API.

Right now, I still want the code to be unittested on this own and we haven't tested fault tolerant on TPU, IPU, horovod yet, so I believe there might be more caviats.

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 7]
```diff
@@ -174,11 +179,18 @@ def on_save_checkpoint(self) -> Dict:
         state: Optional[MergedIteratorState] = getattr(self._data_fetcher.dataloader_iter, state_to_save, None)
         if state:
             state_dict["dataloader_state_dict"] = asdict(state)
+        state_dict["dataloader_state_dict"] = _collect_states_on_rank_zero_over_collection(
+            state_dict["dataloader_state_dict"]
+        )
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:184]: nit: access the dictionary just once
```suggestion
        if state:
            state = asdict(state)
        state = _collect_states_on_rank_zero_over_collection(state)
        state_dict["dataloader_state_dict"] = state
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 8]
```diff
@@ -320,9 +321,14 @@ def on_save_checkpoint(self) -> Dict:
...
         ):
             return state_dict
+
         state_dict["dataloader_state_dict"] = self.trainer.train_dataloader.state_dict(
             has_completed=self._has_completed()
         )
+
+        state_dict["dataloader_state_dict"] = _collect_states_on_rank_zero_over_collection(
+            state_dict["dataloader_state_dict"]
+        )
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:331]: Same comment about avoiding 3 dictionary accesses

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -2094,7 +2094,11 @@ def _results(self) -> Optional[ResultCollection]:
...
+            should_terminate_gracefully = self.training_type_plugin.broadcast(self._terminate_gracefully)
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2099]: shouldn't this use `reduce_boolean_decision`?

File: tests/utilities/test_distributed.py [hunk 10]
```diff
@@ -64,15 +64,14 @@ def foo():
 def _test_collect_states(rank, world_size):
     os.environ["MASTER_ADDR"] = "localhost"
 
+    torch.cuda.set_device(f"cuda:{rank}")
+
```
Comment (carmocca) [inline tests/utilities/test_distributed.py:68]: Why is this required now?

**Review by carmocca — APPROVED** [review]

File: tests/utilities/test_distributed.py (hunk 10 above)
Comment (tchaton) [reply tests/utilities/test_distributed.py:68]: I believe there is a bug with `torch.distributed.broadcast_object_list` or it might be a design choice.

But [here](https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1855) they rely on torch.cuda.current_device() instead of torch.distributed.get_rank() meaning both tensors are still on `cuda:0` and the broadcast would fail.

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:184]: Done !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:331]: Done !

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Actually, no because the signal can be set to 1, many or all processes. Used reduce instead.

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -2094,14 +2094,17 @@ def _results(self) -> Optional[ResultCollection]:
...
-            raise ExitGracefullyException(f"Exiting gracefully on {class_name}:{caller.function}")
+        if not _fault_tolerant_training():
+            return
+        if not self._should_terminated_gracefully():
+            return
+        caller = inspect.stack()[1]
+        class_name = caller[0].f_locals["self"].__class__.__name__
+        raise ExitGracefullyException(f"Exiting gracefully on {class_name}:{caller.function}")
+
+    def _should_terminated_gracefully(self) -> bool:
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2105]:
```suggestion
    def _should_terminate_gracefully(self) -> bool:
```

File: pytorch_lightning/trainer/trainer.py [hunk 12]
```diff
@@ -2094,14 +2094,17 @@ def _results(self) -> Optional[ResultCollection]:
...
-            caller = inspect.stack()[1]
-            class_name = caller[0].f_locals["self"].__class__.__name__
-            raise ExitGracefullyException(f"Exiting gracefully on {class_name}:{caller.function}")
+        if not _fault_tolerant_training():
+            return
+        if not self._should_terminated_gracefully():
+            return
+        caller = inspect.stack()[1]
+        class_name = caller[0].f_locals["self"].__class__.__name__
+        raise ExitGracefullyException(f"Exiting gracefully on {class_name}:{caller.function}")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2103]: I strongly suggest not doing this, this is what stacktraces are for.

File: pytorch_lightning/trainer/trainer.py [hunk 13]
```diff
@@ -2094,10 +2094,17 @@ def _results(self) -> Optional[ResultCollection]:
...
 
     def _exit_gracefully_on_signal(self) -> None:
-        if _fault_tolerant_training() and self._terminate_gracefully:
-            caller = inspect.stack()[1]
-            class_name = caller[0].f_locals["self"].__class__.__name__
-            raise ExitGracefullyException(f"Exiting gracefully on {class_name}:{caller.function}")
+        if not _fault_tolerant_training():
+            return
+        if not self._should_terminated_gracefully():
+            return
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2100]:
```suggestion
        if not _fault_tolerant_training() or not self._should_terminated_gracefully():
            return
```

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:2105]: Done !

File: pytorch_lightning/trainer/trainer.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:2100]: No, you don't want to force a reduce when fault tolerant is not active. This would be costly for the users who don't want to opt-in.
